### PR TITLE
Store non sponsored podcasts from Discover page in the databse

### PR DESCRIPTION
**Pull request by MiSikora**
## Description

This adds periodic partial Discover page caching. At the application startup a periodic worker with a period of 1 day is registered to crawl the Discover page.

I decided not to normalize the `curated_podcasts` table. It would only needlessly complicate things and there isn't a lot of data to be saved.

### DB migration

1. Install app from the b9eb449c98190c032c17a2dd9e972c079bcb47c8.
2. Sign in.
3. Install app from this PR.
4. You shouldn't experience any data loss.

### Syncing

1. Open the app.
2. Wait a bit.
3. Inspect `curated_podcasts` table. It should contain the Feature, Trending, and all non sponsored lists.

## Checklist
- [x] ~If this is a user-facing change, I have added an entry in CHANGELOG.md~
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] ~All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`~
- [x] ~Any jetpack compose components I added or changed are covered by compose previews~
- [x] ~I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.~
 
#### I have tested any UI changes...

- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>CuratedPodcastsSyncWorker</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/db/AppDatabase.kt [hunk 1]
```diff
@@ -830,6 +830,32 @@ abstract class AppDatabase : RoomDatabase() {
...
+                            podcast_id TEXT NOT NULL,
+                            podcast_title TEXT NOT NULL,
+                            podcast_description TEXT,
+                            PRIMARY KEY (list_id, podcast_id)
+                        )
+                        """.trimIndent(),
+                    )
+                    database.execSQL("CREATE INDEX curated_podcasts_list_id_index ON curated_podcasts(list_id)")
+                    database.execSQL("CREATE INDEX curated_podcasts_podcast_id_index ON curated_podcasts(podcast_id)")
+                    database.execSQL("DROP TABLE IF EXISTS trending_podcasts")
```
Comment (MiSikora) [inline modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/db/AppDatabase.kt:851]:
```suggestion
                    database.execSQL("DROP TABLE trending_podcasts")
```

File: modules/features/discover/src/test/kotlin/au/com/shiftyjelly/pocketcasts/discover/worker/CuratedPodcastsCrawlerTest.kt [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+            |    "source": "${server.url("/")}",
+            |    "summary_style": "",
+            |    "expanded_style": "",
+            |    "regions": []
+            |  }
+            |]
+        """.trimMargin("|"),
+        )
+        server.enqueue(MockResponse().setSocketPolicy(SocketPolicy.SHUTDOWN_INPUT_AT_END))
+        server.enqueue(MockResponse().setSocketPolicy(SocketPolicy.SHUTDOWN_INPUT_AT_END))
```
Comment (ashiagr) [inline modules/features/discover/src/test/kotlin/au/com/shiftyjelly/pocketcasts/discover/worker/CuratedPodcastsCrawlerTest.kt:311]: Duplicate line
```suggestion
```

**Review by ashiagr — APPROVED** [review]
Nice work!

**Comment by ashiagr** [comment]
I haven't checked yet, does it impact automotive or wear apps?

File: modules/features/discover/src/test/kotlin/au/com/shiftyjelly/pocketcasts/discover/worker/CuratedPodcastsCrawlerTest.kt (hunk 2 above)
Comment (MiSikora) [reply modules/features/discover/src/test/kotlin/au/com/shiftyjelly/pocketcasts/discover/worker/CuratedPodcastsCrawlerTest.kt:311]: This is correct. An enqueued response is consumed from the server.

**Comment by MiSikora** [comment]
> I haven't checked yet, does it impact automotive or wear apps?

No, the worker is registered only in the phone app.

**Pull request merged**